### PR TITLE
Fix support for next version of @material-ui/core

### PR DESCRIPTION
**Pull request by mnajdova**
Changes done:
- replaces `makeStyles`, `withStyles`, `createStyles` from `@material-ui/core/styles` with ones from `@material-ui/styles`. `defaultTheme` is required to be passed in order to preserve previous behavior.
- ~created `createTheme` util that maps to the legacy `createMuiTheme` or the new `createTheme`~ - reverted these changes see https://github.com/iamhosseindhv/notistack/pull/387#issuecomment-854790479

Not sure how to test the changes locally, any help is appreciated.

Related to https://github.com/mui-org/material-ui/pull/26382

### CODE REVIEW

**Comment by xiaoyu-tamu** [comment]
@mnajdova hi, MUI maintainer. Do you have any ideas why custom jss instance or createGenerateClassName in `<StyleProvider />` does not work well with notistack? 

#289

File: src/SnackbarContainer.tsx
Comment (yurtaev) [inline src/SnackbarContainer.tsx:7]: Wrong path:

```diff
- import createTheme from '../utils/createTheme'
+ import createTheme from './utils/createTheme'
```

File: src/SnackbarItem/SnackbarItem.tsx
Comment (yurtaev) [inline src/SnackbarItem/SnackbarItem.tsx:4]: Should it be `utils/createTheme.js` instead of `createMuiTheme`?

Comment (mnajdova) [reply src/SnackbarItem/SnackbarItem.tsx:4]: Good catch!

File: src/SnackbarContainer.tsx
Comment (mnajdova) [reply src/SnackbarContainer.tsx:7]:
```suggestion
import createTheme from './utils/createTheme';
```

**Comment by mnajdova** [comment]
> @mnajdova hi, MUI maintainer. Do you have any ideas why custom jss instance or createGenerateClassName in <StyleProvider /> does not work well with notistack?

Not sure without diving too deep. Would be great to test whether v5 will have problems like this, after `notistick` is compatible with it.

File: src/SnackbarItem/SnackbarItem.tsx
Comment (mnajdova) [reply src/SnackbarItem/SnackbarItem.tsx:4]:
```suggestion
import { Theme, emphasize } from '@material-ui/core/styles';
import createTheme from '../utils/createTheme';
```

**Comment by mlake** [comment]
Hi @mnajdova  - i'm happy to help test your code in my own project...

in general should be something like the following
```
git clone git@github.com:mnajdova/notistack.git 
cd notistack
git checkout feat/core-styles-dependencies-updates
npm install
npm run build
npm link
```
Then cd to a different project set up with material ui v5...

```
npm link notistack
```

**Comment by mnajdova** [comment]
Thanks @mlake, will try this tomorrow.

**Review by duongdev — APPROVED** [review]

**Comment by oliviertassinari** [comment]
It's probably worth saying that this approach was already tested and released on the data grid with https://github.com/mui-org/material-ui-x/pull/1627

**Comment by mnajdova** [comment]
I've tried testing locally, but there is a problem with the improts. 

Basically the import:

```
import * as styles from '@material-ui/core/styles';
```

is being transformed to:
```
import { createMuiTheme, createTheme as createTheme$1, emphasize } from '@material-ui/core/styles';
``` 

which then fails as `createTheme` is not exported from v4. I would revert this to always use the `createMuiTheme` (it is deprecated in v5, but I don't think there is anything else I can do right now.

**Comment by mnajdova** [comment]
I tried testing locally, but I got some `Invalid hook call.` errors, probably didn't link the project correctly, I am not sure. If someone can test a simple scenario whether it work with both v4 and v5 it would be great.

**Comment by mlake** [comment]
also got the invalid hook call when I tested..wasn't sure how to go about fixing it

**Comment by oliviertassinari** [comment]
@iamhosseindhv Do you have guidance on these changes?

**Comment by DmitriySot** [comment]
looks like you guys also have no idea from which package you need to import

**Comment by oliviertassinari** [comment]
@DmitriySot What do you mean?

**Comment by DmitriySot** [comment]
I mean that material-ui updated a few weeks ago with some breaking changes related to migration and removing jss. And for me is not too clear from which package I need to import something that will not be deprecated in the next version.And I see that you also can't update notistack for that time

**Comment by oliviertassinari** [comment]
> I need to import something that will not be deprecated in the next version

@DmitriySot `@material-ui/styles` will be deprecated and removed.

**Comment by yurtaev** [comment]
I'm just trying to upgrade `@material-ui` to v5.0.0-alpha.35 on my project. I use `yarn patch` as workaround to fix notistack. What about releasing the fix as alpha tag? and pin `@material-ui/*` deps to v5

**Comment by mnajdova** [comment]
> looks like you guys also have no idea from which package you need to import

@DmitriySot comments like this are not really helpful and productive.

> I mean that material-ui updated a few weeks ago with some breaking changes related to migration and removing jss. And for me is not too clear from which package I need to import something that will not be deprecated in the next version. And I see that @mnajdova also can't update notistack for that time

We would like to contribute to make notistack compatible with both Material-UI v4 and v5 at the same time, so that it won't break until it is upgraded to v5. Regarding what should we imported from where, we try to list everything in the migration guide, for these specific changes, there is more info https://github.com/mui-org/material-ui/blob/next/docs/src/pages/guides/migration-v4/migration-v4.md#material-uicorestyles Anyway I don't think it's productive to discuss things like this in this PR, you can open an issue for clarifying things.

Again, whoever can help test out notistack with both v4 and v5 would help with merging the PR.

**Comment by CAIOHOBORGHI** [comment]
I'm getting an error with the new release of MUI.

styles$2.withStyles is not a function
    at Object.<anonymous> (node_modules\notistack\dist\notistack.cjs.development.js:227:47)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (node_modules\notistack\dist\index.js:7:20)

**Comment by yurtaev** [comment]
I prepared codesanboxes:

- [notistack + @material-ui v5](https://codesandbox.io/s/notistack-material-ui-v5-2tuqu?file=/App.js)
- [notistack + @material-ui v4](https://codesandbox.io/s/notistack-material-ui-v4-go8hy?file=/App.js)

I see the warning for v5:

```
Material-UI: the createMuiTheme function was renamed to createTheme.

You should use `import { createTheme } from '@material-ui/core/styles'`
```

**Comment by hauptrolle** [comment]
@yurtaev Hey, it looks like the spacing between the notifications is missing? 🤔

**Comment by belayaShlyapa** [comment]
@iamhosseindhv Do we have some news please ?
I'm still stuck by `Attempted import error: 'makeStyles' is not exported from '@material-ui/core/styles'.` error using material-ui v5

**Comment by mnajdova** [comment]
> I prepared codesanboxes:
> 
> notistack + @material-ui v5
> notistack + @material-ui v4
> I see the warning for v5:
> 
> Material-UI: the createMuiTheme function was renamed to createTheme.
> 
> You should use `import { createTheme } from '@material-ui/core/styles'`

This is great, thanks @yurtaev Once notistack is switched to v5, we can rename the import from `createMuiTheme` to `createTheme`, in order to fix the warning. Also once we switch to v5, we could switch to using `styled()` instead of `makeStyles`/`withStyles`

**Comment by MichaelKheel** [comment]
> I prepared codesanboxes:
> 
> * [notistack + @material-ui v5](https://codesandbox.io/s/notistack-material-ui-v5-2tuqu?file=/App.js)
> * [notistack + @material-ui v4](https://codesandbox.io/s/notistack-material-ui-v4-go8hy?file=/App.js)
> 
> I see the warning for v5:
> 
> ```
> Material-UI: the createMuiTheme function was renamed to createTheme.
> 
> You should use `import { createTheme } from '@material-ui/core/styles'`
> ```

Thanks for fixing it, it really works.

**Comment by lukaszmysiak** [comment]
> 
> 
> @yurtaev Hey, it looks like the spacing between the notifications is missing? 🤔

@hauptrolle There is a change in Collapse element global class / rule naming API, i.e. v4's `.MuiCollapse-container` > `.MuiCollapse-root` in v5.
It seems that there was a change in 1.0.8 which dropped margin based spacing in favor of paddings on `.MuiCollapse-wrapper` which is selected by v4 css API:
```
const collapse = {
    container: '& > .MuiCollapse-container',
    wrapper: '& > .MuiCollapse-container > .MuiCollapse-wrapper',
};
```
https://github.com/iamhosseindhv/notistack/blob/master/src/SnackbarContainer.tsx#L8-L9

**Comment by benediktdertinger** [comment]
Hi guys, we are receiving the same error as @CAIOHOBORGHI  does: `TypeError: styles$2.withStyles is not a function`

We are on:
```json
"@emotion/react": "^11.4.0",
"@emotion/styled": "^11.3.0",
"@material-ui/core": "^5.0.0-alpha.37",
"@material-ui/icons": "^4.11.2",
...
"next": "^11.0.0",
...
"react": "^17.0.2",
"react-dom": "^17.0.2",
```

I already tried `yarn add notistack` + `yarn add notistack@next` but the problem still remains. I also checked out the provided codesandboxes but they use a different material UI version and we depend on the lastet one. Do you have any suggestions?

**Comment by MichaelKheel** [comment]
> Hi guys, we are receiving the same error as @CAIOHOBORGHI does: `TypeError: styles$2.withStyles is not a function`
> 
> We are on:
> 
> ```json
> "@emotion/react": "^11.4.0",
> "@emotion/styled": "^11.3.0",
> "@material-ui/core": "^5.0.0-alpha.37",
> "@material-ui/icons": "^4.11.2",
> ...
> "next": "^11.0.0",
> ...
> "react": "^17.0.2",
> "react-dom": "^17.0.2",
> ```
> 
> I already tried `yarn add notistack` + `yarn add notistack@next` but the problem still remains. I also checked out the provided codesandboxes but they use a different material UI version and we depend on the lastet one. Do you have any suggestions?

install `notistack-1.0.9-next.tgz` from @yurtaev

**Comment by benediktdertinger** [comment]
> > Hi guys, we are receiving the same error as @CAIOHOBORGHI does: `TypeError: styles$2.withStyles is not a function`
> > We are on:
> > ```json
> > "@emotion/react": "^11.4.0",
> > "@emotion/styled": "^11.3.0",
> > "@material-ui/core": "^5.0.0-alpha.37",
> > "@material-ui/icons": "^4.11.2",
> > ...
> > "next": "^11.0.0",
> > ...
> > "react": "^17.0.2",
> > "react-dom": "^17.0.2",
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I already tried `yarn add notistack` + `yarn add notistack@next` but the problem still remains. I also checked out the provided codesandboxes but they use a different material UI version and we depend on the lastet one. Do you have any suggestions?
> 
> install `notistack-1.0.9-next.tgz` from @yurtaev

Thank you. We implemented it with the patched version and the basics work BUT NOT the additional action passed to the snackbar item. It is also not working in the example from @yurtaev. You can test it [here](https://codesandbox.io/s/notistack-material-ui-v5-2tuqu?file=/MessageButtons.js) by clicking on `DEFAULT` and then `My Action`. It seems that somehow the element is not accepting events at all as no mouse cursor, hover state etc. are displayed...

**Comment by kjellski** [comment]
I've read through this thread and I'm currently not able to help as I'd need guidance on what's missing. I'd be interested to do so though, has anyone enough information on what the missing pieces are? 
I'm also trying to use 

```
        "notistack": "1.0.6-next.0",
        "@material-ui/core": "^5.0.0-alpha.37",
```

currently - which fails with the following after install:

```
./node_modules/notistack/dist/notistack.esm.js
Attempted import error: 'makeStyles' is not exported from '@material-ui/core/styles'.
```

**Comment by benediktdertinger** [comment]
> @material-ui/core/styles

@kjellski you have to also install `yarn add @material-ui/core/styles`. After I did, it worked.

**Review by stass-1 — APPROVED** [review]

**Review by MichaelKheel — APPROVED** [review]

**Comment by Hummel37** [comment]
@iamhosseindhv 
The beta of material-ui is now out. Can this pr be merged?

**Comment by xiaoyu-tamu** [comment]
@iamhosseindhv

**Comment by yurtaev** [comment]
Note: related #396 to fix https://github.com/iamhosseindhv/notistack/pull/387#issuecomment-863149122

**Comment by deldrid1** [comment]
When `yarn install`ing  with a `dependency` key/value pair of  `"notistack": "https://github.com/mnajdova/notistack.git#82a1975b255b8be6b43b3d61c03bac36366eae47",
` in `package.json`, I get the following error emitted:

`warning " > notistack@1.0.9" has incorrect peer dependency "@material-ui/core@^4.0.0".`

On top of that something appears to still be broken - My `next.js` app complains with 
- `Module not found: Can't resolve 'notistack'`
- `>  import { SnackbarProvider } from 'notistack';`

Any suggestions on how to use the MUI v5 beta with this?  The Code Sandbox example above appears to play nice...

**Comment by iamhosseindhv** [comment]
Apologies for responding so late. Been dealing with Covid and international travel. Couple of things I can think of @mnajdova:
1. Do we have to list `@material-ui/styles` in peerDependencies now?
2. If so, I wouldn't merge this into master as it'd force users that are locked with MUI v4 to download the new dependency. Currently the branch structure is as follows:

| branch  | Published with tag | Compatible |
| ------------- | ------------- |------------- |
| `master`  | `latest`  | MUI v4  |
| `next`  | `next`  |  MUI v5  |

**Comment by Neuromancer93** [comment]
thanks @iamhosseindhv . I look forward to trying out the new version. Hope all is well.

**Comment by lpfrenette** [comment]
I think that material-ui made the decision for you. I ran into the problem that changes in 4.12 reflect was will be happening in 5.x. So i think you master branch should switch to 4.12 or 5 and above and people who locked their version of material-ui to 4.11 and bellow most likely have also lock notistack to 1.09. If not they will update material ui or lock notistack to 1.09.

Anyway you know what your doing! 
Thanks for doing it!

**Comment by mwhitney-butter** [comment]
`"@material-ui/core": "^5.0.0-beta.0"`
`"notistack": "^1.0.6-next.1"`

```
Failed to compile.

./node_modules/notistack/dist/notistack.esm.js
Attempted import error: 'makeStyles' is not exported from '@material-ui/core/styles'.
```

thoughts?

**Comment by ivanvladimirov** [comment]
> 
> 
> `"@material-ui/core": "^5.0.0-beta.0"`
> `"notistack": "^1.0.6-next.1"`
> 
> ```
> Failed to compile.
> 
> ./node_modules/notistack/dist/notistack.esm.js
> Attempted import error: 'makeStyles' is not exported from '@material-ui/core/styles'.
> ```
> 
> thoughts?

That's what this PR is about to fix

**Comment by mwhitney-butter** [comment]
> > `"@material-ui/core": "^5.0.0-beta.0"`
> > `"notistack": "^1.0.6-next.1"`
> > ```
> > Failed to compile.
> > 
> > ./node_modules/notistack/dist/notistack.esm.js
> > Attempted import error: 'makeStyles' is not exported from '@material-ui/core/styles'.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > thoughts?
> 
> That's what this PR is about to fix

what's the timeline for this PR being merged?

**Comment by mnajdova** [comment]
> Could this PR be rebased and merged into next branch?

I thought there is one branch (master) to support all versions. If this is not the case, I would say the best is to migrate to emotion in the `next` branch so that there isn't any overhead added with bundling both JSS and emotion.

**Comment by mwhitney-butter** [comment]
need a tagged version of notistack on master that uses material-ui v4 with JSS and a tagged version of notistack on master that uses material-ui v5 with emotion so people can choose which version they want. until notistack next version gets merged into master, it should only have emotion, then when next version gets merged into master, the release will be the aforementioned tag that uses material-ui v5.

> > Could this PR be rebased and merged into next branch?
> 
> I thought there is one branch (master) to support all versions. If this is not the case, I would say the best is to migrate to emotion in the `next` branch so that there isn't any overhead added with bundling both JSS and emotion.

**Comment by mtr1990** [comment]
When will we have a version that supports MUI beta v5?

I've been waiting for this package to update for 2 months but still no results.

**Comment by ivanvladimirov** [comment]
> When will we have a version that supports MUI beta v5?
> 
> I've been waiting for this package to update for 2 months but still no results.

Same here. Good thing my project is still in development. I hope this PR gets merged soon.

**Comment by Hummel37** [comment]
Just upgrade to notistack@next and it will work with material UI 5

**Comment by vbornand** [comment]
> Just upgrade to notistack@next and it will work with material UI 5

It's not totally correct, it works only with MUI >= 5.00-alpha.34:
> * Unstable MUI `>= 5.0.0-alpha.35`:  Wait for [Fix support for next version of @material-ui/core #387](https://github.com/iamhosseindhv/notistack/pull/387) to be merged.
> * MUI `<= 5.0.0-alpha.34` or `>= 4.12.0`:  Use `npm i notistack@next` or lock to version `npm i notistack@1.0.6-next.1`. See a working [example](https://codesandbox.io/s/epic-star-dzx7k?file=/App.js)
> * MUI `< 4.12.0`: Use the latest. `npm i notistack` or `npm i notistack@latest`.

**Comment by ivanvladimirov** [comment]
Material-UI is currently at its second beta of v5 and styling functions are not part of "core" anymore, therefore notistack cannot work with the new version of MUI

**Comment by mnajdova** [comment]
@iamhosseindhv I propose updating to the `styled()` and `sx` API in the next branch, so that it doesn't depend on two styling solutions (JSS and emotion). 

When this PR was introduced, the goal was to make notistack compatible with both v4 and v5 in the transition period. Now that Material-UI is in beta and this was still not merged, I propose we abandon it and focus on updating the `next` branch. Let me know if you need some help with/instructions with it.

**Comment by stass-1** [comment]
@iamhosseindhv currently we don't have any opened tickets or pull-requests to track progress with getting notistack work with MUI v5 beta

**Comment by iamhosseindhv** [comment]
@stass-1 https://github.com/iamhosseindhv/notistack/issues/412 is an open issue to make notistack compatible with MUI 5 beta

**Comment by th851dan** [comment]
mui5 is released and has a new package name(@mui/material). Do you want to make notistack compatible? @iamhosseindhv

**Comment by mnajdova** [comment]
> mui5 is released and has a new package name(@mui/material). Do you want to make notistack compatible? @iamhosseindhv

There is a codemod that should enable fast migration to the new package names - https://github.com/mui-org/material-ui/blob/next/packages/mui-codemod/README.md#mui-replace

**Comment by iamhosseindhv** [comment]
Thanks @mnajdova. @th851dan We have #423 for that issue

**Pull request closed without merging**
